### PR TITLE
DEVPROD-8993 Expand module names when cloning with git.get_project

### PR DESCRIPTION
**Pull request by ZackarySantana**
DEVPROD-8993

### Description
This expands the module name (if it should be expanded).

### Testing
Add a unit test that uses an expansion named module

### Documentation
Mention in the docs about how it can expand

### CODE REVIEW

**Review by ablack12 — COMMENTED** [review]
I think this also needs to be considered here https://github.com/evergreen-ci/evergreen/blob/7138c86e543fffa80d9706e542c1b3ae4564b4d8/agent/command/git.go#L1065 

I'm technically okay with this change but I think that it's really likely we're going to miss things here when we make future changes. Do we currently allow for expansions to be used in build variant or task definitions? I think expansions are really meant to be used only in command definitions (but obviously we do have some exceptions, like module prefix) so we should maybe be careful about what opening this up means. Are there any alternatives for the user here?

File: agent/command/git.go [hunk 1]
```diff
@@ -866,6 +866,10 @@ func (c *gitFetchProject) fetch(ctx context.Context,
 
 	// For every module, expand the module prefix.
```
[the remark this answers is not in the document]
Comment (ZackarySantana) [reply agent/command/git.go:867]: Hm that makes sense. I thought about what might differentiate this between other build variant fields and I think it's that the modules section is a bit of an anti-pattern.

We've seen this with the correctness ticket, that the git.get_project attempts to clone more repo's than is listed in the build variant definition (since they re-used the same func/git.get_project across all build_variant's). The modules section is _only_ tied to the git.get_project command while every other field ties in to most/every stage of a task. 

I think in an ideal world, build variants wouldn't list the modules and the git.get_project's would only. How it's set up now, the modules section of the build variant are really just param's for git.get_project. 

I'm okay if we do push back on this, since it is something that we generally can't support with any other build variant fields, but I do think it's different enough (and if we wanted to, could make a mini-project of migrating the fields) to warrant an exception. I could increase the wording in the docs / our code to exemplify that.

File: docs/decisions/2024-07-11_include_file_limits.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# 2024-03-26 Appropriate limit for include files per project
+# 2024-07-11 Appropriate limit for include files per project
```
Comment (ZackarySantana) [inline docs/decisions/2024-07-11_include_file_limits.md:1]: I noticed that the date in the header and the date in the file name was different than the date of the ADR, I can undo this in this ticket but thought might as well shape it up

File: agent/command/git.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply agent/command/git.go:867]: I think you're right that modules are something of an exception since we allow expansions for that section already; I do think we should clarify that this is an exception in the docs so I'll leave a comment to clarify.

File: docs/decisions/2024-07-18_allow_module_expansions.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+- status: approved
+- date: 2024-07-18
+- authors: Zackary Santana
+
+## Context and Problem Statement
+
+Ticket: DEVPROD-8993
+
+Typically, all project commands use options from their `params` field to configure the command- allowing the command to be customizable at runtime. The `git.get_project` command uses the `modules` field from the build variant the command is running in, which is not expandable and not customizable at runtime.
```
Comment (ablack12) [inline docs/decisions/2024-07-18_allow_module_expansions.md:11]: Well stated problem statement!

File: docs/decisions/2024-07-18_allow_module_expansions.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+2. Create a new parameter in the `git.get_project` command that overrides the `modules` field in the build variant configuration. This field would be expandable but the `modules` field in the build variant configuration would remain unexpandable.
+
+3. Recommend users generate a build variant with `generate.tasks` that includes the module they want to use in the `git.get_project` command. This would allow the user to customize the module at runtime.
+
+## Decision Outcome
+
+Option 2 introduces a new flow for the command which would make it harder to debug. Right now, the workflow for debugging and handling `git.get_project` command has been the same. We don't currently have a pattern where complex configuration is overridden by a command parameter so this option is preferred. Option 3 was rejected because it introduces extra steps, more resources, and more complexity for the user and it breaks the customizability of commands.
```
[the remark this answers is not in the document]
Comment (ablack12) [reply docs/decisions/2024-07-18_allow_module_expansions.md:23]: Could you explain "and it breaks the customizability of commands"?

[the remark this answers is not in the document]
Comment (ZackarySantana) [reply docs/decisions/2024-07-18_allow_module_expansions.md]: essentially, if commands could not expand values, user's would do the workflow of option 3 (generate.tasks to make the commands they want with hard coded values). This concept of generating commands with hard coded values goes against the customizability of commands, expansions, and parameters. 

I will reword this to: `Option 3 was rejected because it introduces extra steps, more resources, and more complexity for the user and goes against making commands customizable for easy & efficient reusability"

File: docs/Project-Configuration/Project-Configuration-Files.md [hunk 5]
```diff
@@ -469,9 +471,9 @@ Fields:
 (note: all fields can be expanded by project variables with the exception of `auto_update`)
 
 -   `name`: alias to refer to the module
--   `branch`: the branch of the module to use in the project
+-   `branch`: the branch of the module to use in the project. These can be [expansions](Project-Configuration-Files#expansions)
 -   `repo`: the git repository of the module
--   `prefix`: the path prefix to use for the module
+-   `prefix`: the path prefix to use for the module. These can be [expansions](Project-Configuration-Files#expansions)
```
Comment (ZackarySantana) [inline docs/Project-Configuration/Project-Configuration-Files.md:476]: I added these expansion links so user's better know what is allowed, I originally was trying to see where we document that the prefix _can_ be an expansion

**Review by ablack12 — APPROVED** [review]
Apologies for the slow review!
